### PR TITLE
[ENG-415][ENG-798] Add periodic task that deactivates requested accounts with tests

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

To speed up the deactivation request process we want to automatically delete users. Now users who request deactivation get emailed or deactivated at midnight each night.

## Changes

- add periodic script and tests
- remove now unnecessary emails sending and throttling tests

## QA Notes

To test this issue you need to have an admin user with permissions that can deactivate an account. Once you do that you can create some test users, make sure some are admin on nodes, both the only admin and shared with other admins. Have though users request deletion and wait until midnight for the period task to run. The task should send our admin emails and deactivate completely those without nodes, preprints, groups etc.

## Documentation

Not user facing, no docs.

## Side Effects

None that I know of.

## Ticket

https://openscience.atlassian.net/browse/ENG-415

### CODE REVIEW

File: website/settings/defaults.py [hunk 1]
```diff
@@ -578,6 +578,10 @@ class CeleryConfig:
                 'task': 'scripts.generate_sitemap',
                 'schedule': crontab(minute=0, hour=5),  # Daily 12:00 a.m.
             },
+            'generate_sitemap': {
```
Comment (brianjgeiger) [inline website/settings/defaults.py]: Drive-by: This should have a different name

File: scripts/periodic/deactivate_requested_accounts.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            user.disable_account()
+            mails.send_mail(
+                to_addr=user.username,
+                mail=mails.REQUEST_DEACTIVATION_COMPLETE,
+                user=user,
+                contact_email=OSF_CONTACT_EMAIL,
+                can_change_preferences=False,
+            )
+
+        user.contacted_deactivation = True
```
Comment (UdayVarkhedkar) [inline scripts/periodic/deactivate_requested_accounts.py]: Why is `contacted_deactivation` set as true even if the user has resources and the email is sent to OSF support?

Comment (Johnetordoff) [reply scripts/periodic/deactivate_requested_accounts.py]: The user still has to be contacted, so they know to contact support to delete their own resources. So either way they are contacted and the bool `contacted_deactivation` is going to be `True`. It would be nice to specify how they were contacted, but this is simpler.

Comment (UdayVarkhedkar) [reply scripts/periodic/deactivate_requested_accounts.py]: It could be beneficial to double check the text of the REQUEST_DEACTIVATION email sent to OSF Support to ensure that it would result in the user being contacted.

File: website/mails/mails.py [hunk 3]
```diff
@@ -251,6 +251,8 @@ def get_english_article(word):
 REQUEST_EXPORT = Mail('support_request', subject='[via OSF] Export Request')
 REQUEST_DEACTIVATION = Mail('support_request', subject='[via OSF] Deactivation Request')
```
Comment (UdayVarkhedkar) [inline website/mails/mails.py]: Based on this, it looks like the email sent to support if user the requests account deletion but has associated resources mentions the Deactivations Request in the subject and provides basic user information (id, profile url, email) via the support_request template. I don't know enough about how support handles these emails to comment on how account deletion is handled after they are contacted.

Comment (Johnetordoff) [reply website/mails/mails.py]: @UdayVarkhedkar  I checked with Product and this is the expected format.

File: osf/models/user.py [hunk 4]
```diff
@@ -1705,6 +1708,15 @@ def gdpr_delete(self):
...
         self.deleted = timezone.now()
 
+    @property
+    def has_resources(self):
+        # TODO: Update once quickfolders in merged
+
+        nodes = self.nodes.exclude(type='osf.quickfilesnode').exists()
+        quickfiles = self.nodes.get(type='osf.quickfilesnode').files.exists()
+
+        return nodes or quickfiles or self.preprints.exists()
```
[the remark this answers is not in the document]
Comment (pattisdr) [reply osf/models/user.py]: Also, deleted preprints aren't excluded - 
something like -  Preprint.objects.filter(_contributors=self, ever_public=True, deleted__isnull=True)

File: osf/models/user.py [hunk 5]
```diff
@@ -1705,6 +1708,15 @@ def gdpr_delete(self):
         self.external_identity = {}
         self.deleted = timezone.now()
 
+    @property
+    def has_resources(self):
```
Comment (pattisdr) [inline osf/models/user.py]: Add a docstring to this please!

File: website/settings/defaults.py [hunk 6]
```diff
@@ -578,6 +579,10 @@ class CeleryConfig:
                 'task': 'scripts.generate_sitemap',
                 'schedule': crontab(minute=0, hour=5),  # Daily 12:00 a.m.
             },
+            'deactivate_requested_accounts': {
+                'task': 'scripts.periodic.deactivate_requested_accounts',
+                'schedule': crontab(minute=0, hour=5),  # Daily 12:00 a.m.
```
Comment (pattisdr) [inline website/settings/defaults.py]: Check with Matt over the best time to run this!

File: scripts/periodic/deactivate_requested_accounts.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+from website.settings import OSF_SUPPORT_EMAIL, OSF_CONTACT_EMAIL
+
+from scripts.utils import add_file_logger
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+
+def deactivate_requested_accounts(dry_run=True):
+    users = OSFUser.objects.filter(requested_deactivation=True, contacted_deactivation=False, date_disabled__isnull=True)
```
Comment (pattisdr) [inline scripts/periodic/deactivate_requested_accounts.py]: Thinking through this...suggestions welcome, but I think with this logic a user can only request deactivation once.  Say I request deactivation and this script runs, which will mark `contacted_deactivation` as True.  Say I have non-deleted projects, so the help desk is emailed.  I then cancel my deactivation request, so Sara doesn't delete my user.   At any point in the future, I won't be able to request again because contacted_deactivation is True.

I'm wondering if a user cancels their deactivation request if we should also set `contacted_deactivation` back to False.

File: website/mails/mails.py [hunk 8]
```diff
@@ -251,6 +251,8 @@ def get_english_article(word):
...
 
+REQUEST_DEACTIVATION_COMPLETE = Mail('request_deactivation_complete', subject='[via OSF] Deactivation Request')
```
Comment (pattisdr) [inline website/mails/mails.py]: I'd recommend changing the subject to something like `OSF account deactivated` instead of using the same subject we send to the help desk?

File: website/settings/defaults.py [hunk 9]
```diff
@@ -472,6 +472,7 @@ class CeleryConfig:
         'scripts.generate_sitemap',
         'scripts.premigrate_created_modified',
         'scripts.add_missing_identifiers_to_preprints',
+        'scripts.periodic.deactivate_requested_accounts',
```
Comment (pattisdr) [inline website/settings/defaults.py]: 👍

File: scripts/periodic/deactivate_requested_accounts.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    for user in users:
+        if user.has_resources:
+            mails.send_mail(
+                to_addr=OSF_SUPPORT_EMAIL,
+                mail=mails.REQUEST_DEACTIVATION,
+                user=user,
+                can_change_preferences=False,
+            )
+        else:
+            user.disable_account()
```
Comment (pattisdr) [inline scripts/periodic/deactivate_requested_accounts.py]: Can you check with product and make sure we're doing everything they're doing when they manually deactivate an account? Wondering if we should additionally mark `is_registered=False` - but check with product first.

File: website/settings/defaults.py (hunk 6 above)
Comment (Johnetordoff) [reply website/settings/defaults.py]: Matt said this is too short to really worry about, so we good!

File: osf/models/user.py [hunk 11]
```diff
@@ -1705,6 +1708,15 @@ def gdpr_delete(self):
...
+        # TODO: Update once quickfolders in merged
+
+        nodes = self.nodes.exclude(type='osf.quickfilesnode').exists()
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply osf/models/user.py]: > Ticket says users cannot be attached to any deleted resources. 

Ticket:
> User cannot be attached to any non-deleted items. 

I'm assuming typo.

[the remark this answers is not in the document]
Comment (pattisdr) [reply osf/models/user.py]: sorry, nondeleted, but yes, if a user is attached to a deleted item, it shouldn't affect them from being automatically disabled

File: website/profile/views.py [hunk 12]
```diff
@@ -837,5 +823,6 @@ def request_deactivation(auth):
 def cancel_request_deactivation(auth):
     user = auth.user
     user.requested_deactivation = False
+    user.contacted_deactivation = False  # In case we've already contacted them once.
```
[the remark this answers is not in the document]
Comment (pattisdr) [reply website/profile/views.py]: We also need to do this in v2, not just v1. v2 -> `api/users/serializers.py`,  `UserSettingsUpdateSerializer.update`.  `requested_deactivation` can be toggled to True or False, so if it's toggled to False, we should flip contacted_deactivation there as well.

File: scripts/periodic/deactivate_requested_accounts.py (hunk 10 above)
Comment (pattisdr) [reply scripts/periodic/deactivate_requested_accounts.py]: Can you confirm that you verified with product that we should set `is_registered` to False?

File: website/settings/defaults.py [hunk 13]
```diff
@@ -486,6 +486,7 @@ class CeleryConfig:
         'scripts.generate_sitemap',
         'scripts.premigrate_created_modified',
         'scripts.add_missing_identifiers_to_preprints',
+        'management.commands.deactivate_requested_accounts',
```
Comment (pattisdr) [inline website/settings/defaults.py]: I think this import needs to be `'osf.management.commands.deactivate_requested_accounts'`.  Can you run the management locally to verify before sending back to CR?

File: osf/management/commands/deactivate_requested_accounts.py [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+        user.contacted_deactivation = True
+        user.email_last_sent = timezone.now()
+        if not dry_run:
+            user.save()
+
+    if dry_run:
+        logger.info('Dry run complete')
+
+
+@celery_app.task(name='management.commands.check_crossref_dois')
```
Comment (pattisdr) [inline osf/management/commands/deactivate_requested_accounts.py]: copy-paste error should not be `check_crossref_dois`!

File: osf/models/user.py [hunk 15]
```diff
@@ -1785,6 +1789,29 @@ def gdpr_delete(self):
...
+        If a user only has no resources or only deleted resources this will return false and they can safely be deactivated
+        otherwise they must delete or transfer their outstanding resources.
+
+        :return bool: does the user have any active node, preprints, groups, quickfiles etc?
+        """
+        from osf.models import Preprint
+
+        # TODO: Update once quickfolders in merged
+
+        nodes = self.nodes.exclude(type='osf.quickfilesnode', is_deleted=True).exists()
```
Comment (pattisdr) [inline osf/models/user.py]: You want to return non-deleted files that are not quickfiles - so this instead? As it is currently, you're returning a queryset of all nodes as long as they are not deleting quickfile nodes.
this instead?
```
self.nodes.exclude(type='osf.quickfilesnode').filter(is_deleted=True)
```

File: osf/migrations/0174_osfuser_contacted_deactivation.py [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+        ('osf', '0173_ensure_schemas'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='osfuser',
+            name='contacted_deactivation',
+            field=models.BooleanField(default=False),
+        ),
+    ]
```
Comment (pattisdr) [inline osf/migrations/0174_osfuser_contacted_deactivation.py]: Out of date with develop - please rebase this migration!

File: osf/management/commands/deactivate_requested_accounts.py [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+                mail=mails.REQUEST_DEACTIVATION_COMPLETE,
+                user=user,
+                contact_email=OSF_CONTACT_EMAIL,
+                can_change_preferences=False,
+            )
+
+        user.contacted_deactivation = True
+        user.email_last_sent = timezone.now()
+        if not dry_run:
+            user.save()
```
Comment (pattisdr) [inline osf/management/commands/deactivate_requested_accounts.py]: I'd wrap more of this in a dry_run - if it's a `dry_run`, you're still sending emails to users, just not saving the information, which could lead to confusion.  I'd recommend reworking so on a dry_run, so no emails are sent, no fields are changed or saved.  Perhaps a dry_run just logs a list of users with resources, and a list without so we can double check these users are correct before running it for real.

File: osf/management/commands/deactivate_requested_accounts.py [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+            '--dry',
+            action='store_true',
+            dest='dry_run',
+            help='Dry run',
+        )
+
+    # Management command handler
+    def handle(self, *args, **options):
+        dry_run = options.get('dry_run', True)
+        main(dry_run=dry_run)
```
Comment (pattisdr) [inline osf/management/commands/deactivate_requested_accounts.py]: Thanks for moving into a management command ⭐️

File: scripts/tests/test_deactivate_requested_accounts.py [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+
+        deactivate_requested_accounts(dry_run=False)
+        user_requested_deactivation_with_node.reload()
+
+        assert user_requested_deactivation_with_node.requested_deactivation
+        assert not user_requested_deactivation_with_node.is_disabled
+        mock_mail.assert_called_with(can_change_preferences=False,
+                                     mail=mails.REQUEST_DEACTIVATION,
+                                     to_addr=settings.OSF_SUPPORT_EMAIL,
+                                     user=user_requested_deactivation_with_node)
```
Comment (pattisdr) [inline scripts/tests/test_deactivate_requested_accounts.py]: Are you sure these tests run? Please run them locally too, they should have failed with your last change, and they did not.

Comment (Johnetordoff) [reply scripts/tests/test_deactivate_requested_accounts.py]: Ugh this is terrible Testmon did not pick up changes in the tests themselves, which includes imports so the coverage info is was wrong. Hmmm, fixing the imports will jog testmon and make this pass, I don't know what to do about it in the future though.

File: osf/management/commands/deactivate_requested_accounts.py [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+
+def deactivate_requested_accounts(dry_run=True):
+    users = OSFUser.objects.filter(requested_deactivation=True, contacted_deactivation=False, date_disabled__isnull=True)
+
+    for user in users:
+        if user.has_resources:
+            logger.info('User {} was contacted for about deactivation.'.format(user._id))
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply osf/management/commands/deactivate_requested_accounts.py]: went with "OSF support is being emailed about deactivating the account of user <guid>"

File: api/users/serializers.py [hunk 21]
```diff
@@ -526,18 +526,13 @@ def verify_two_factor(self, instance, value, two_factor_addon):
...
-                    mail=mails.REQUEST_DEACTIVATION,
-                    user=instance,
-                    can_change_preferences=False,
-                )
-                instance.email_last_sent = timezone.now()
             instance.requested_deactivation = requested_deactivation
-            instance.save()
-        return
+
+        instance.save()
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply api/users/serializers.py]: What about a situation like what [this is testing for](https://github.com/CenterForOpenScience/osf.io/blob/develop/api_tests/users/views/test_user_settings_detail.py#L270-L277) ? if `contacted_deactivation` just happens to be true and we want to deactivate we have to toggle it again. 

That said I don't think `contacted_deactivation` can just happen to be true if `requested_deactivation` is False, so I'll just fix that test case.
